File: src/cleansia_android/core/src/main/java/cz/cleansia/core/consent/SignupConsentStore.kt
package cz.cleansia.core.consent

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.core.stringSetPreferencesKey
import javax.inject.Inject
import javax.inject.Qualifier
import javax.inject.Singleton
import kotlinx.coroutines.flow.first

/**
 * Qualifies the preferences `DataStore` backing [SignupConsentStore]. Each app provides
 * one under its own file name, the same arrangement
 * [cz.cleansia.core.notifications.PushTokenDataStore] uses.
 */
@Qualifier
@Retention(AnnotationRetention.BINARY)
annotation class SignupConsentDataStore

/**
 * Persists the tick across the process death that can sit between signup and the first
 * sign-in — the user reads the confirmation mail on another device and comes back a day
 * later. Survives sign-out on purpose: the tick belongs to the account, not the session.
 */
@Singleton
class SignupConsentStore @Inject constructor(
    @SignupConsentDataStore private val dataStore: DataStore<Preferences>,
) {

    suspend fun save(pending: PendingSignupConsent) {
        dataStore.edit { prefs ->
            prefs[KEY_EMAIL] = pending.email
            prefs[KEY_TYPES] = pending.types.map(SignupConsentType::name).toSet()
        }
    }

    suspend fun read(): PendingSignupConsent? {
        val prefs = dataStore.data.first()
        val email = prefs[KEY_EMAIL]?.takeIf { it.isNotBlank() } ?: return null
        val names = prefs[KEY_TYPES].orEmpty()
        val types = SignupConsentType.entries.filter { it.name in names }
        return if (types.isEmpty()) null else PendingSignupConsent(email, types)
    }

    suspend fun settle(type: SignupConsentType) {
        dataStore.edit { prefs ->
            val remaining = prefs[KEY_TYPES].orEmpty() - type.name
            if (remaining.isEmpty()) {
                prefs.remove(KEY_EMAIL)
                prefs.remove(KEY_TYPES)
            } else {
                prefs[KEY_TYPES] = remaining
            }
        }
    }

    private companion object {
        val KEY_EMAIL = stringPreferencesKey("pending_email")
        val KEY_TYPES = stringSetPreferencesKey("pending_types")
    }
}

File: src/cleansia_ios/CleansiaCore/Sources/CleansiaCore/Consent/SignupConsentStore.swift
import Foundation

public protocol SignupConsentStore: AnyObject, Sendable {
    func save(_ pending: PendingSignupConsent)
    func read() -> PendingSignupConsent?
    func settle(_ type: SignupConsentType)
}

/// Persists the tick across the process death that can sit between signup and the first
/// sign-in — the user reads the confirmation mail on another device and comes back a day
/// later. Deliberately outside the session-wipe registry: the tick belongs to the account,
/// not the session, and the address it is keyed on is what stops it reaching another one.
public final class UserDefaultsSignupConsentStore: SignupConsentStore, @unchecked Sendable {
    private enum Key {
        static let email = "consent.pending_email"
        static let types = "consent.pending_types"
    }

    private let defaults: UserDefaults

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    public func save(_ pending: PendingSignupConsent) {
        defaults.set(pending.email, forKey: Key.email)
        defaults.set(pending.types.map(\.rawValue), forKey: Key.types)
    }

    public func read() -> PendingSignupConsent? {
        guard let email = defaults.string(forKey: Key.email), !email.isEmpty else { return nil }
        let types = storedTypes()
        return types.isEmpty ? nil : PendingSignupConsent(email: email, types: types)
    }

    public func settle(_ type: SignupConsentType) {
        let remaining = storedTypes().filter { $0 != type }
        if remaining.isEmpty {
            defaults.removeObject(forKey: Key.email)
            defaults.removeObject(forKey: Key.types)
        } else {
            defaults.set(remaining.map(\.rawValue), forKey: Key.types)
        }
    }

    private func storedTypes() -> [SignupConsentType] {
        let raw = defaults.array(forKey: Key.types) as? [Int] ?? []
        return raw.compactMap(SignupConsentType.init(rawValue:))
    }
}
